aspnetcore/blazor/components/prerender.md

```diff
@@ -17,358 +17,134 @@ uid: blazor/components/prerender
-  * `RenderMode.Server`: The service is available for the Interactive Server render mode.
-  * `RenderMode.Webassembly`: The service is available for the Interactive Webassembly render mode.
-  * `RenderMode.InteractiveAuto`: The service is available for both the Interactive Server and Interactive Webassembly render modes if a component renders in either of those modes.
-* The service is resolved during the initialization of an interactive render mode, and the properties annotated with the `[SupplyParameterFromPersistentComponentState]` attribute are deserialized.
-
-> [!NOTE]
-> Only persisting scoped services is supported.
-
-<!-- UPDATE 10.0 - Flesh out with a fully-working example. -->
+@page "/"
+@inject IWebAssemblyHostEnvironment Environment
 
-`CounterService.cs`:
+<PageTitle>Home</PageTitle>
 
-```csharp
-public class CounterService
-{
-    [SupplyParameterFromPersistentComponentState]
-    public int CurrentCount { get; set; }
+<h1>Home</h1>
 
-    public void IncrementCount()
-    {
-        CurrentCount++;
-    }
-}
+<p>
+    Environment: @Environment.Environment
+</p>
 ```
 
-In `Program.cs`:
-
-```csharp
-builder.Services.RegisterPersistentService<CounterService>(
-    RenderMode.InteractiveAuto);
-```
-
-Serialized properties are identified from the actual service instance:
-
-* This approach allows marking an abstraction as a persistent service.
-* Enables actual implementations to be internal or different types.
-* Supports shared code in different assemblies.
-* Results in each instance exposing the same properties.
+No compile time error occurs, but a runtime error occurs during prerendering:
 
-As an alternative to using the declarative model for persisting state with the `[SupplyParameterFromPersistentComponentState]` attribute, you can use the <xref:Microsoft.AspNetCore.Components.PersistentComponentState> service directly, which offers greater flexibility for complex state persistence scenarios. Call <xref:Microsoft.AspNetCore.Components.PersistentComponentState.RegisterOnPersisting%2A?displayProperty=nameWithType> to register a callback to persist the component state during prerendering. The state is retrieved when the component renders interactively. Make the call at the end of initialization code in order to avoid a potential race condition during app shutdown.
+> :::no-loc text="Cannot provide a value for property 'Environment' on type 'BlazorSample.Client.Pages.Home'. There is no registered service of type 'Microsoft.AspNetCore.Components.WebAssembly.Hosting.IWebAssemblyHostEnvironment'.":::
 
-The following counter component example persists counter state during prerendering and retrieves the state to initialize the component.
+This error occurs because the component must compile and execute on the server during prerendering, but <xref:Microsoft.AspNetCore.Components.WebAssembly.Hosting.IWebAssemblyHostEnvironment> isn't a registered service on the server.
```

Comment:
Yeah, it's a bit annoying that we don't have a common service interface for getting the environment that works on both the server and client.